model.py

```diff
@@ -239,24 +256,153 @@ def decode(
+        input_size = h_encoder.shape[0]
+        model_out = output = torch.zeros(
+            [1, batch_size, self.embedding.weight.shape[1]]
+        ).to(self.embedding.weight.device)
+        hidden = None
+        device = self.embedding.weight.device
+        decoded_idxs = (
+            torch.LongTensor([bos_idx]).repeat(batch_size).unsqueeze(0).to(device)
+        )
+        eos_generated = np.zeros((1, batch_size), dtype=np.bool)
+
+        init_state = {
+            "idxs": decoded_idxs,
+            "hidden": hidden,
+            "attention": torch.zeros([0, batch_size, input_size]).to(device),
+            "model_out": model_out,
+            "score": torch.zeros(batch_size).to(device),
+            "is_done": np.zeros(batch_size, dtype=bool),
+        }
+        states: List[Dict[str, Any]] = [init_state]
+        keep_going = True
+        iters = 0
+        final_states = []
+        while len(states) > 0 and iters < max_decoding_len:
+            candidates: List[Dict] = []
+            for state in states:
+                candidates += self._next_states(state, h_encoder, beam_size)
+            states = self._filter_top_k(candidates, beam_size)
+            keep_going = False
+            i = 0
+            while i < len(states):
+                if states[i]["is_done"].all():
+                    final_states.append(states.pop(i))
+                else:
+                    i += 1
+            iters += 1
+        final_states += states
+
+        top_idxs = torch.cat([s["score"] for s in final_states], 0).argmax(0)
+        return [final_states[ci]["idxs"][:, bi] for bi, ci in enumerate(top_idxs)]
+
+    def _filter_top_k(self, candidates, k) -> List[Dict]:
+        scores = torch.cat([c["score"] for c in candidates], 0)
+        top_idxs = torch.topk(scores, k, dim=0).indices
+        dict_slice = lambda k: [c[k] for c in candidates]
+        states: List[Dict] = []
+        for group_idxs in top_idxs:
+            new_idxs: List[Tensor] = []  # n, bs
+            new_hidden0: List[Tensor] = []  # 2, bs, 1024
+            new_hidden1: List[Tensor] = []  # 2, bs, 1024
+            new_attention: List[Tensor] = []  # n, bs, len_in
+            new_model_out: List[Tensor] = []  # 1, bs, 300
+            new_score: List[Tensor] = []  # 1, bs
+            new_is_done: List[Tensor] = []  # 1, bs (numpy)
+            # Batch index, candidate index
+            for bi, ci in enumerate(group_idxs):
+                new_idxs.append(candidates[ci]["idxs"][:, bi])
+                new_hidden0.append(candidates[ci]["hidden"][0][:, bi, :])
+                new_hidden1.append(candidates[ci]["hidden"][1][:, bi, :])
+                new_attention.append(candidates[ci]["attention"][:, bi, :])
+                new_model_out.append(candidates[ci]["model_out"][:, bi, :])
+                new_score.append(candidates[ci]["score"][:, bi])
+                new_is_done.append(candidates[ci]["is_done"][bi])
+            states.append(
```

Comment:
I think if we ensure everything has a uniform(ish) shape, we can turn this into something reasonable.